### DISCUSSION TITLE
Where to place webhook listeners?

### DISCUSSION THREAD

**Discussion by rdtclark** [pay-rails/pay#486] [Q&A] [answered]
I am following the docs here https://github.com/pay-rails/pay/blob/master/docs/stripe/1_overview.md regarding creating a webhook listener to save order details to the database.

I understand I need to add the following to an initializer file e.g.

```Ruby
# config/initializer/pay.rb
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", FulfillCheckout.new
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.async_payment_succeeded", FulfillCheckout.new
```

But I am confused about where to place the listener itself??

```Ruby
# app/models/order/webhooks/fulfill_checkout.rb??
class FulfillCheckout
  def call(event)
    object = event.data.object

    if object.payment_status == "paid"
      # example
      order = Order.create!(address: object.charges.shipping.address, user_id: object.charges.data.metadata.user_id)
      order.paid!
    end
  end
end
```

**Comment by excid3** [comment, marked as answer]
You can define the class anywhere you want. I like putting mine in `app/webhooks` but you can put it in models, lib, etc.

**Reply by nhocki** [reply]
Hi @excid3 I'm trying this but my class is not being found (`NameError: uninitialized constant Foo`). It won't find _any_ class defined in `app/`. Do you think this might be because of new loader in Rails 7?

In the meantime I solved this by doing, but not sure if there's a way to prevent the extra lambda.

```ruby
Pay::Webhooks.delegator.subscribe "stripe.charge.succeeded", ->(event) { Foo.new.call(event) }
```

Thanks!

**Comment by mpvosseller** [comment]
@excid3 Curious how you would namespace webhooks at the application level?

- `app/webhooks/pay/stripe/charge_succeeded.rb` (`Pay::Stripe::ChargeSucceeded`) - Using pay's namespace `Pay::` seems bad.
- `app/webhooks/stripe/charge_succeeded.rb` (`Stripe::ChargeSucceeded`) - Using stripe's namespace `Stripe::` seems bad.
- `app/webhooks/charge_succeeded.rb` (`ChargeSucceeded`) - No namespace wouldn't scale to additional processors.
- `app/webhooks/stripe_charge_succeeded.rb` (`StripeChargeSucceeded`) - Prepending the processor name doesn't seem sustainable.

**Reply by nhocki** [reply]
@mpvosseller fwiw we ended up using `Payments::Stripe::Webhooks::SubscriptionCreated`

**Reply by mpvosseller** [reply]
@nhocki Good idea. And out of curiosity where do you put your `payments` directory? (`app/payments`, `app/lib/payments`, `app/services/payments`, etc..). Thanks!

**Reply by excid3** [reply]
You could do `app/pay/webhooks/charge_succeeded.rb`

**Reply by nhocki** [reply]
We have it in `lib/payments` but `app/payments` might be better 🤔
